### PR TITLE
py-build: new versions 0.10.0 (-> flit)

### PR DESCRIPTION
**Pull request by wdconinc**
py-build switched to flit with 0.10, https://github.com/pypa/build/blob/0.10.0/pyproject.toml

```
==> py-build: Successfully installed py-build-0.10.0-twgngkplyegllaovlp45r76nsk7bqezw
```

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-build/package.py [hunk 1]
```diff
@@ -12,15 +12,21 @@ class PyBuild(PythonPackage):
...
+    depends_on("py-flit-core@3.4:", when="@0.10:", type="build")
+    depends_on("py-setuptools", when="@:0.9", type="build")
     depends_on("py-packaging@19:", type=("build", "run"))
-    depends_on("py-pep517@0.9.1:", type=("build", "run"))
+    depends_on("py-pep517@0.9.1:", when="@:0.9", type=("build", "run"))
     depends_on("py-tomli@1:", type=("build", "run"))
+    depends_on("py-tomli@1.1:", when="@0.10:", type=("build", "run"))
     depends_on("py-colorama", when="platform=windows", type=("build", "run"))
     depends_on("py-importlib-metadata@0.22:", when="^python@:3.7", type=("build", "run"))
     depends_on("py-virtualenv@20.0.35:", when="+virtualenv", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-build/package.py:33]: Can you check to see if this patch is still needed and update if so?

File: var/spack/repos/builtin/packages/py-build/package.py [hunk 2]
```diff
@@ -12,15 +12,21 @@ class PyBuild(PythonPackage):
...
     variant("virtualenv", default=False, description="Install optional virtualenv dependency")
 
     depends_on("python@3.6:", type=("build", "run"))
-    depends_on("py-setuptools", type="build")
+    depends_on("python@3.7:", when="@0.10:", type=("build", "run"))
+    depends_on("py-flit-core@3.4:", when="@0.10:", type="build")
+    depends_on("py-setuptools", when="@:0.9", type="build")
     depends_on("py-packaging@19:", type=("build", "run"))
-    depends_on("py-pep517@0.9.1:", type=("build", "run"))
+    depends_on("py-pep517@0.9.1:", when="@:0.9", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-build/package.py:28]: Missing dep on py-pyproject-hooks

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-build/package.py (hunk 2 above)
Comment (wdconinc) [reply var/spack/repos/builtin/packages/py-build/package.py:28]: `py-pyproject-hooks` package and dependency added. Thanks to https://github.com/spack/spack/pull/35326/files#diff-60a30aca51f9b417b47bf3689c36fa9d65e469f076e6e1035a28e94bdb10318a for some of the `package.py` (but not using bdist wheel).

File: var/spack/repos/builtin/packages/py-pyproject-hooks/package.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+from spack.package import *
+
+
+class PyPyprojectHooks(PythonPackage):
+    """Wrappers to call pyproject.toml-based build backend hooks."""
+
+    homepage = "https://github.com/pypa/pyproject-hooks"
+    url = "https://github.com/pypa/pyproject-hooks/archive/refs/tags/v1.0.0.tar.gz"
+    list_url = "https://github.com/pypa/pyproject-hooks/tags/"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyproject-hooks/package.py]: https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi-vs-github

File: var/spack/repos/builtin/packages/py-build/package.py (hunk 1 above)
Comment (wdconinc) [reply var/spack/repos/builtin/packages/py-build/package.py:33]: Ah, that again. I recall the reason for that patch now, but not exactly how to trigger the issue.

In any case:
- the current patch, introduced in #28562 to address [pypa/build #266](https://github.com/pypa/build/issues/266), does not apply to any other versions,
- there were multiple PRs addressing [pypa/build #266](https://github.com/pypa/build/issues/266) that were opened/closed with lots of discussion, but the issue itself remains open.

File: var/spack/repos/builtin/packages/py-pyproject-hooks/package.py (hunk 3 above)
Comment (wdconinc) [inline var/spack/repos/builtin/packages/py-pyproject-hooks/package.py]:
```suggestion
    pypi = "pyproject_hooks/pyproject_hooks-1.0.0.tar.gz"
```

File: var/spack/repos/builtin/packages/py-pyproject-hooks/package.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+    version("1.0.0", sha256="d45c52f9af6bce94755eecf9dbfe6b3c89ef9a50088a8809f5bbec4ed0f9be0b")
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/py-pyproject-hooks/package.py]:
```suggestion
    version("1.0.0", sha256="f271b298b97f5955d53fb12b72c1fb1948c22c1a6b70b315c54cedaca0264ef5")
```

File: var/spack/repos/builtin/packages/py-build/package.py (hunk 1 above)
Comment (wdconinc) [reply var/spack/repos/builtin/packages/py-build/package.py:33]: I can't find any reason to think this has been fixed. I'll extend the patch for all versions as a filter_file.

**Review by adamjstewart — APPROVED** [review]
Confirmed that it installs on my system and allows me to build sdists and bdists of my projects.

**Pull request merged**